### ISSUE TITLE
VAE KL Divergence

### ISSUE THREAD

**Issue by nmichlo** [PyTorchLightning/lightning-bolts#565]
The current VAE implementation uses the following to calculate the KL divergence, which depends on the sampled `z` value.

This seems to differ from the original VAE and BetaVAE papers?

https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/1bec12200ff60b9b58277b40d9232933e1aa9dbe/pl_bolts/models/autoencoders/basic_vae/basic_vae_module.py#L135-L140

I am curious as to why the KL divergence is not directly computed from the posterior `q` and prior `p` distributions themselves?

```python
kl = torch.distributions.kl_divergence(q, p)
kl = kl.mean()
kl *= self.kl_coeff 
```

**Comment by akihironitta** [comment]
@nmichlo Hi, thank you for reporting the issue!

If I understand it correctly, this is about whether to analytically compute KL or to approximate it with `z` sampled from the posterior distribution q. I think this is doing the same thing basically, but I believe we should compute KL between the posterior q and prior p, as you pointed out, because the KL can be analytically computed.

Would you be interested in submitting a PR to fix this?

**Comment by nmichlo** [comment]
I understand the reasoning now behind that approach.

I can submit a PR for the changes. Although I am not sure if the choice of "approximate" or "analytical" methods should be added as a hyperparameter, or if the goal is just to keep the VAE module as simple as possible?

**Comment by akihironitta** [comment]
@nmichlo Thank you for having interests in contributing! As far as I know, there is no reason for approximately computing KL where it's possible to compute it analytically. (Please correct me if I'm misunderstanding.) So, shall we just replace the current KL with the analytical one?

**Comment by nmichlo** [comment]
As far as I understand that is correct. I was about to put together a PR when I noticed #566, since it relates it might be easier combing these issues in a single PR?

----------

I also notice that the [bolts VAE docs](https://pytorch-lightning-bolts.readthedocs.io/en/latest/autoencoders.html?highlight=vae#basic-vae) state that you can override the distributions, however this functionality is currently not present in the VAE class. Other distributions may have different parameterisations and require encoder changes so it might be worth adjusting this example and adding the required methods to the VAE class? I am not sure if I should submit another PR?

**Comment by akihironitta** [comment]
> As far as I understand that is correct. I was about to put together a PR when I noticed #566, since it relates it might be easier combing these issues in a single PR?

This issue can be, I think, resolved independently of #566, so would you submit a PR addressing only this issue?

> I also notice that the bolts VAE docs state that you can override the distributions, however this functionality is currently not present in the VAE class. Other distributions may have different parameterisations and require encoder changes so it might be worth adjusting this example and adding the required methods to the VAE class? I am not sure if I should submit another PR?

Would you mind creating an issue about it so that we can discuss it there?
